### PR TITLE
generalize upsampling

### PR DESCRIPTION
**Pull request by mo271**

### CODE REVIEW

File: jxl/src/render/stages/upsample.rs [hunk 1]
```diff
@@ -3,24 +3,35 @@
...
-pub struct Upsample2x {
-    kernel: [[[[f32; 5]; 5]; 2]; 2],
+pub struct Upsample<const N: usize, const SHIFT: u8> {
+    kernel: [[[[f32; 5]; 5]; N]; N],
     channel: usize,
 }
 
-impl Upsample2x {
+impl<const N: usize, const SHIFT: u8> Upsample<N, SHIFT> {
+    // TODO(firsching): remove once we use this!
```
Comment (veluca93) [inline jxl/src/render/stages/upsample.rs:17]: you can `#![allow(dead_code)]` at the beginning of the file to have this apply to the whole file

File: jxl/src/render/stages/upsample.rs [hunk 2]
```diff
@@ -3,24 +3,35 @@
...
     #[allow(dead_code)]
-    pub fn new(ups_factors: &CustomTransformData, channel: usize) -> Upsample2x {
-        let weights = ups_factors.weights2;
-        let n = 1;
-        let mut kernel = [[[[0.0; 5]; 5]; 2]; 2];
+    pub fn new(ups_factors: &CustomTransformData, channel: usize) -> Self {
+        assert_eq!(
```
Comment (veluca93) [inline jxl/src/render/stages/upsample.rs:20]: to ensure compile-time eval, you can...

```
const { assert!(1 << SHIFT == N) }
```

**Review by veluca93 — APPROVED** [review]

File: jxl/src/render/stages/upsample.rs (hunk 1 above)
Comment (mo271) [reply jxl/src/render/stages/upsample.rs:17]: done

File: jxl/src/render/stages/upsample.rs (hunk 2 above)
Comment (mo271) [reply jxl/src/render/stages/upsample.rs:20]: done

**Pull request merged**